### ISSUE TITLE
Feature Request: Ability to merge JSON output files

### ISSUE THREAD

**Issue by michael1026** [pry0cc/axiom#155]
When running axiom-scan, there's some great utilities for merging different types of files, but currently, JSON is not supported. Not sure if this would be a heavy lift, but as the default (or sometimes optional) output format of many tools, it'd be great to be able to merge all of the JSON output files into one final JSON file that's downloaded in the end. I don't think this would be too difficult with Python.

**Comment by pry0cc** [comment]
Which tools have json output you'd like to merge?

**Comment by michael1026** [comment]
Most recently, I was thinking of writing a utility to setup Arjun, which outputs as a JSON file.

**Comment by 0xtavian** [comment]
@michael1026 we added Arjun! instead of JSON we output as txt. This just made it easier to add to axiom without having to write code to merge JSON. But we still plan to do that sometime in the future.

**Comment by 0xtavian** [comment]
@michael1026 Just to review this issue, im not sure if merging arbitrary JSON will every be a thing an axiom. there are too many cavorts ( as far as I can tell ). The main issue is how to merge. If merging by common JSON fields that might be problematic and lead to undesirable results. There seem to be too many variation between JSON structures to support merging them.

One solution that I've used for a while, with nuclei ( or any PD tool for that matter ), when you output as json, they store the entire json object as one line. Currently just changing the nuclei/httpx/dnsx module command in axiom to output as json ( but still keeping the "ext" as "txt", works really well for JSON. 

Using this workaround for all tools that support outputting to json could possibly work. we might be able to get away with first force all JOSN output to one line per result, then cat-ing everything into one file and simply sort -u. Something like this might work for tools that dont support outputting json to one line. Basically force output to one line and then merge the same as we do with txt extensions. 

` find $tmp/output/ -type f -exec cat {} \; | jq -c | sort -u `
 LMK any thoughts.

**Comment by pry0cc** [comment]
I agree. Honestly, the txt option with a directory is the best solution IMO.

That being said, merging JSON is an interesting idea. But true, it hasn't dawned on me that there is an obvious way to merge schema, nor, should it be in scope. 

I would be interested to here if Michael has an idea on a specific use case where this is needed or desirable that I'm missing?

Thanks!

**Comment by michael1026** [comment]
Hey,

I've been getting around this by outputting JSON into indiviual files, then merging them using `sed`. Here's what I'm currently using...

```
sed -i '/{}/d' myFile.json
sed -i ':a;N;$!ba;s/}\n{/,/g' myFile.json
```

This has worked perfectly fine for the tools I have, which outputs like so...

```
{"https://www.example.com/":{"params":["user"]}
{"https://www.yahoo.com/":{"params":["page"]}
```

I completely understand where you're coming from though. In situations where (like most JSON), the keys aren't unique, you would have to merge keys, which becomes complex. As convenient as it would be, it's probably not worth the struggle :) 

Thanks!

**Comment by 0xKourama** [comment]
I have a problem how i can edit arjun model to make each input line a file  output inside a directory

**Comment by gilillo32** [comment]
> @michael1026 
> 
> One solution that I've used for a while, with nuclei ( or any PD tool for that matter ), when you output as json, they store the entire json object as one line. Currently just changing the nuclei/httpx/dnsx module command in axiom to output as json ( but still keeping the "ext" as "txt", works really well for JSON.

Hello, when you mention that you output the nuclei results as json. How do you do it if axiom only allow txt? Thanks in advance.

**Comment by mdpung** [comment]
Also came here looking for support around json. Output from `httpx` supports output in json format, which is very useful when adding additional probes for information. I would expect for every json object that gets returned to at least have a comma. The only workaround is to continue the output in json from `httpx`, but then specify `txt` in the module.

These are the commands I apply afterwards to cleanup the json:
```sh
$ sed '$!s/$/,/' yourfile.json # Adds a comma after each line, for json object separation (except the last)
$ sed -i '1s/^/[/' yourfile.json # Adds an opening bracket at the very beginning
$ echo ']' >> yourfile.json # Adds a closing bracket at the very end
```

It seems there could use for supporting custom extensions or custom logic. Obviously defining how to merge arbitrary json is complicated, and everyone seems to have their own requirements. Since these can be fixed with just a few bash commands, what if we can define exactly that in the module json. I'll open up an issue for the proposal of this.

**Comment by mdpung** [comment]
> Also came here looking for support around json. Output from `httpx` supports output in json format, which is very useful when adding additional probes for information. I would expect for every json object that gets returned to at least have a comma. The only workaround is to continue the output in json from `httpx`, but then specify `txt` in the module.
> 
> These are the commands I apply afterwards to cleanup the json:
> 
> $ sed '$!s/$/,/' yourfile.json # Adds a comma after each line, for json object separation (except the last)
> $ sed -i '1s/^/[/' yourfile.json # Adds an opening bracket at the very beginning
> $ echo ']' >> yourfile.json # Adds a closing bracket at the very end
> 
> It seems there could use for supporting custom extensions or custom logic. Obviously defining how to merge arbitrary json is complicated, and everyone seems to have their own requirements. Since these can be fixed with just a few bash commands, what if we can define exactly that in the module json. I'll open up an issue for the proposal of this.

I've added a feature request here https://github.com/pry0cc/axiom/issues/825.
